src/control/server/mgmt_pool_test.go

```diff
@@ -625,29 +628,121 @@ func TestServer_MgmtSvc_PoolDestroy(t *testing.T) {
 			req:    &mgmtpb.PoolDestroyReq{},
 			expErr: errors.New("empty pool id"),
 		},
-		"fails due to engine error": {
+		// Note: evict dRPC fails, still expect a PoolDestroyResp from PoolDestroy
+		"evict dRPC fails with -DER_BUSY due to open handles force=false, pool still ready": {
+			req: &mgmtpb.PoolDestroyReq{Id: mockUUID},
+			expDrpcEvReq: &mgmtpb.PoolEvictReq{
+				Sys:          build.DefaultSystemName,
+				Id:           mockUUID,
+				SvcRanks:     []uint32{0, 1, 2},
+				Destroy:      true,
+				ForceDestroy: false,
+			},
+			setupMockDrpc: func(svc *mgmtSvc, err error) {
+				setupMockDrpcClient(svc, &mgmtpb.PoolDestroyResp{
+					Status: int32(drpc.DaosBusy),
+				}, nil)
+			},
+			expResp: &mgmtpb.PoolDestroyResp{
+				Status: int32(drpc.DaosBusy),
+			},
+			expSvcState: stateAddr(system.PoolServiceStateReady),
+		},
+		"evict dRPC fails due to engine error": {
+			req: &mgmtpb.PoolDestroyReq{Id: mockUUID},
+			expDrpcEvReq: &mgmtpb.PoolEvictReq{
+				Sys:          build.DefaultSystemName,
+				Id:           mockUUID,
+				SvcRanks:     []uint32{0, 1, 2},
+				Destroy:      true,
+				ForceDestroy: false,
+			},
+			setupMockDrpc: func(svc *mgmtSvc, err error) {
+				setupMockDrpcClient(svc, &mgmtpb.PoolDestroyResp{
+					Status: int32(drpc.DaosMiscError),
+				}, nil)
+			},
+			expResp: &mgmtpb.PoolDestroyResp{
+				Status: int32(drpc.DaosMiscError),
+			},
+			expSvcState: stateAddr(system.PoolServiceStateDestroying),
+		},
+		"force=true, evict dRPC fails due to engine error": {
+			req: &mgmtpb.PoolDestroyReq{Id: mockUUID, Force: true},
+			expDrpcEvReq: &mgmtpb.PoolEvictReq{
+				Sys:          build.DefaultSystemName,
+				Id:           mockUUID,
+				SvcRanks:     []uint32{0, 1, 2},
+				Destroy:      true,
+				ForceDestroy: true,
+			},
+			setupMockDrpc: func(svc *mgmtSvc, err error) {
+				setupMockDrpcClient(svc, &mgmtpb.PoolDestroyResp{
+					Status: int32(drpc.DaosMiscError),
+				}, nil)
```

Comment:
I'm a little confused by some of these cases. We're setting up a PoolDestroyResp as the response, but the PoolEvictResp isn't set up. The setupMockDrpcClient only sets up one response, which it mocks over and over, so I guess maybe PoolDestroyResp and PoolEvictResp look the same to protobuf when unmarshaling.

Even if it technically works I think this is destined to be confusing. If you don't want to change it now, we should make a ticket to change it later. What I'm thinking is that we should change setupMockDrpcClient to accept a slice of proto.Message objects, so it can offer different types of responses.

Reply:
This probably needs another ticket / PR to add new testing flexibility. It was a little confusing to make even this change, being less familiar with this testing code structure and realizing a function under test (PoolDestroy) that invokes 1 (or 2) dRPCs of different kinds may need further enhancements.

Yes, if setupMockDrpcClient could accept multiple messages then this would enable certain scenarios that cannot be currently tested (evict dRPC success, destroy dRPC failure).

How would the test code here take advantage of setting up a PoolEvictResp if the PoolDestroy function returns PoolDestroyResp? I made the PoolDestroy() implementation always return a PoolDestroyResp even if it only invoked the evict dRPC internally. Maybe there is more flexibility somehow to inspect the responses of the individual component dRPCS?